Add requests per second graph to the HTML report.

slashrsm opened the pull request:
Closes #246.

## TODO

- ~Graph color scheme~
- ~Only display starting/stopping phase when `--no-reset-metrics` is used~
- Investigate spike on transition between starting and started phase
- Confirm that metrics are correct (when throttling is enabled RPS numbers do not match exactly)
- Automated tests
- Add additional graphs. 

## Screenshots

### With `--no-reset-metrics`
<img width="1041" alt="Screenshot 2021-11-18 at 20 53 10" src="https://user-images.githubusercontent.com/376889/142495119-2002196c-8cf3-44eb-ab7e-92f8f74242b9.png">


### Without `--no-reset-metrics`
<img width="1066" alt="Screenshot 2021-11-18 at 20 49 50" src="https://user-images.githubusercontent.com/376889/142495134-ed00e11c-1241-4620-854c-1308761ffc52.png">

slashrsm commented on src/config.rs:
```diff
@@ -68,6 +68,8 @@ const DEFAULT_PORT: &str = "5115";
 /// --debug-format FORMAT      Sets debug log format (csv, json, raw, pretty)
 /// --no-debug-body            Do not include the response body in the debug log
 /// --status-codes             Tracks additional status code metrics
+/// --requests-per-second      Tracks additional requests per second metrics
+/// --rps-bucket               Size (in seconds) of the time-based requests per second bucket (default: 10)
```
Any ideas for better names?

slashrsm replied:
These are now gone and so is the concept of "bucket".

slashrsm commented on src/metrics.rs:
```diff
@@ -451,6 +458,26 @@ impl GooseRequestMetricAggregate {
+
+    /// Record request in requests per second metric.
+    pub(crate) fn record_requests_per_second(&mut self, bucket_size: usize) {
+        let time_bucket = (Utc::now().timestamp() - self.start_time) as usize / bucket_size;
```
Instead of the actual time of the request we are using the time when the request metric is reported. 

It would be better to collect the time when the request happened, but I think that this data currently isn't reported by the worker.

slashrsm commented on src/report.rs:
```diff
@@ -404,6 +406,46 @@ pub fn error_row(error: &metrics::GooseErrorMetricAggregate) -> String {
+        <h2>Requests per second</h2>
+            <div id="main" style="width: 1000px; height:660px; background: white;"></div>
+
+            <script src="https://cdn.jsdelivr.net/npm/echarts@5.2.2/dist/echarts.min.js"></script>
```
Is it OK to load echarts from the CDN?Is it OK to load echarts from the CDN?

LionsAd replied:
That's fine to start - we could make this configurable in the future - if you wanted to host your own.

slashrsm replied:
Great points. Mentioned the library in the docs.

jeremyandrews commented on src/metrics.rs:
```diff
@@ -2951,6 +2987,33 @@ impl GooseAttack {
 
+            let graph_rps_template: String;
+            if self.configuration.requests_per_second {
+                let mut aggregated_count: HashMap<usize, usize> = HashMap::new();
```
I wonder if instead we could use the logger thread to efficiently write time-data to log files (I assume we'd need multiple: one for each type of chart we add to the reports). So instead of buckets, in the main parent loop we'd track the total # of requests, total # of tasks, total # of users seen each second and push that to the logger thread. Then, when we write the reports we could perhaps read in the time-data from these files instead of trying to hold them in memory? In this way we always have access to per-second granularity, but we can decide how often to plot a point (every second, every 10 seconds, every minute, etc) depending on how long the load test actually ran, something we can't know at start-time and run-time), and we can even add controls allow us to zoom in/out on the data.

slashrsm replied:
I changed the approach and now it stores # of requests for each second as you described here. 

I calculated how much data that approach uses in memory and I got about ~350kB per day (4 bytes/second * 86400 = 345600 bytes). Considering how little memory this uses I am not sure if it is worth adding the complexity with storing data to files. Thoughts?

jeremyandrews replied:
Completely agreed, there's no reason to add this sort of complexity for such small data sizes -- even if we add many charts.

jeremyandrews reviewed (changes requested):
It's fantastic seeing this feature come together! 

As we only use the requests-per-second (and tasks-per-second and users-per-second etc) data when writing the html-report, perhaps we simply tie it to the existing `--report-file` option.

Specifying a bucket size doesn't seem to scale well, we need a mechanism that works regardless of if the load test runs for 5 minutes or 5 hours.

The code is clean and when it works it looks great -- very nice addition to the load test report!

slashrsm commented on src/metrics.rs:
```diff
@@ -9,6 +9,7 @@
 //! [`GooseErrorMetrics`] are displayed in tables.
 
 use chrono::prelude::*;
+use chrono::Utc;
```
Removed.

slashrsm commented on src/config.rs:
```diff
@@ -319,6 +329,10 @@ pub(crate) struct GooseDefaults {
     /// An optional default to track additional status code metrics.
     pub status_codes: Option<bool>,
+    /// An optional default to track requests per second metrics.
+    pub requests_per_second: Option<bool>,
```
This is a great point. I removed other options and tied the graph to the report being enabled. 

We can always add more granular control at a later point if it turns out that we need it.

slashrsm commented on src/metrics.rs:
```diff
@@ -396,6 +397,10 @@ pub struct GooseRequestMetricAggregate {
     /// that all Workers are running the same load test plan.
     pub load_test_hash: u64,
+    // Timestamp when the metric collection was started.
+    start_time: i64,
```
This is now gone. I switched to `GooseMetrics.duration`.

slashrsm commented on src/config.rs:
```diff
@@ -1495,6 +1509,42 @@ impl GooseConfiguration {
+                    message: "rps_bucket",
+                },
+            ])
+            .unwrap_or(10);
```
This is now gone.

slashrsm commented:
I removed the options that were added in the initial version of this pull request and now we rely on `--report-file` to enable/disable this.

I also changed how the data is stored. Now we are storing the number of requests for each second and the (flawed) concept of "buckets" is now gone. Let me know what you think. 

Also, what do you think about using something like this for the chart style: https://echarts.apache.org/examples/en/editor.html?c=area-simple. I imagine that it could be quite useful specially for long-running load tests (ability to zoom in, etc.).

jeremyandrews commented on src/config.rs:
```diff
@@ -187,10 +187,8 @@ pub struct GooseConfiguration {
     #[options(no_short)]
     pub no_debug_body: bool,
     /// Tracks additional status code metrics
-    // Add a blank line and then an Advanced: header after this option
```
Remove this and the next unnecessary change in this file.

slashrsm replied:
Done.

jeremyandrews commented on src/metrics.rs:
```diff
@@ -2415,6 +2434,9 @@ impl GooseAttack {
                 merge_request.set_status_code(request_metric.status_code);
             }
+            if !self.configuration.report_file.is_empty() {
+                merge_request.record_requests_per_second(self.metrics.duration);
```
We already have `self`, so no need to pass `self.metrics.duration`

jeremyandrews replied:
The graph should include ramp-up time and ramp-down time -- in the following test I ran with the following parameters:

```bash
cargo run --release --example drupal_memcache -- --host http://apache/ -v -u5 -t10 --report-file report.html --throttle-requests 11
```

It's not including the ramp-up time, and it needs to. Also, I think we're missing the first second, as we should see 10 seconds of requests for ~11 rps:

<img width="1115" alt="image" src="https://user-images.githubusercontent.com/402892/141246254-b88e9eb9-dedf-46ae-adf4-312d5bf4b620.png">

slashrsm replied:
> We already have self, so no need to pass self.metrics.duration

`record_requests_per_second()` is a function on a different struct so `self` represents  something else there. And `duration` is not available there, so we need to send it.

jeremyandrews reviewed (changes requested):
This is much cleaner, thanks for reworking! 

The graph should include ramp-up, running, and ramp-down (and ideally should visually indicate each somehow). I also like the idea of a different graph type that allows you to zoom in.

jeremyandrews commented:
> I also changed how the data is stored. Now we are storing the number of requests for each second and the (flawed) concept of "buckets" is now gone. Let me know what you think.

This is much simpler, I like it!

Note that the data is still getting cut off: I ran a load test for 1 hour (3600 seconds) but the data was cut off -- I manually added the missing bit of the report to see the data that was recorded. Likely this is a limitation of the current `--report-file` implementation, and a buffer or something will need to be increase to support larger file sizes.

<img width="1130" alt="image" src="https://user-images.githubusercontent.com/402892/141265783-5c3eb5e4-6b90-4418-8385-91156822b0ba.png">

> Also, what do you think about using something like this for the chart style: https://echarts.apache.org/examples/en/editor.html?c=area-simple. I imagine that it could be quite useful specially for long-running load tests (ability to zoom in, etc.).

I like that idea very much, it would add richness to the charts being able to zoom in.

Perhaps the X axis should be actual timestamps though, instead of seconds since the test started? I wonder which is more useful? (The latter would be easier to match up against data in New Relic etc)

slashrsm commented:
I am still trying to figure out the cutting off problem (I am able to reproduce it). I fell that there is something wrong with JSON serialization or maybe `format!` when the data series grows over certain point. Looking into it as we speak, but is quite time consuming, since I can only reproduce it if I run a load test for a while.

slashrsm commented on src/report.rs:
```diff
@@ -404,6 +406,48 @@ pub fn error_row(error: &metrics::GooseErrorMetricAggregate) -> String {
+                    }},
+                    series: [
+                        {{
+                            data: {values},
```
This was caused by using `write` instead of `write_all`. My runs are producing valid HTML now.

slashrsm commented:
I improved the appearance of the graph:

1. Improved styling of the line
2. Added zoom support
3. Added tooltip on mouse hover
4. Marked starting and stopping period (I can't make the latter to appear as it completes so quickly)
5. Added "Save as image" function.

I noticed that there is always a spike of requests exactly at the point when the load test was started up and `--no-reset-metrics`. Is it possible that we are recording some requests twice in this case?

## Without `--no-reset-metrics`
![Screenshot 2021-11-11 at 20-32-39 Goose Attack Report](https://user-images.githubusercontent.com/376889/141365383-00306856-427b-4191-995b-7289551f62a5.png)

## With `--no-reset-metrics`
![Screenshot 2021-11-11 at 20-38-36 Goose Attack Report](https://user-images.githubusercontent.com/376889/141365781-77a0910f-8084-4932-9964-27ac3629a9cf.png)

jeremyandrews commented:
I think your graphs in the PR are labeled backwards -- the first one looks like `--no-reset-metrics` IS enabled, the second looks like `--no-reset-metrics` is NOT enabled.

jeremyandrews commented:
Any idea what this spike is at the start of the load test? It seems to show up in the graphs semi-regularly (in this case I'm using the following options: `-v --report-file report.html -u10 -s10 -t10 --no-reset-metrics`:

<img width="921" alt="image" src="https://user-images.githubusercontent.com/402892/141456819-c620f760-f872-48c4-b323-b3648b5d493e.png">

jeremyandrews commented:
Any idea what this spike is at the start of the load test? It seems to show up in the graphs semi-regularly (in this case I'm using the following options: `-v --report-file report.html -u10 -s10 -t10 --no-reset-metrics`:

<img width="921" alt="image" src="https://user-images.githubusercontent.com/402892/141456819-c620f760-f872-48c4-b323-b3648b5d493e.png">

jeremyandrews commented on src/metrics.rs:
```diff
@@ -3004,6 +3030,44 @@ impl GooseAttack {
+                && Some(stopped) == self.metrics.stopped
+            {
+                let mut count = 0;
+                for (_path, path_metric) in self.metrics.requests.iter() {
```
If `--report-file` is enabled and `--no-reset-metrics` is not enabled, we should store a copy of the `metrics.requests` "somewhere else" before we reset the metrics so we can generate a complete chart. Alternatively, in this configuration perhaps we should simply not display the "Starting" portion of the chart as it will always be blank.

jeremyandrews reviewed (changes requested):
This is looking very nice! I like the new charts much better.
 - When we are displaying starting/stopping, perhaps we could use "gray" instead of a red shade, as the latter draws more attention to it than it likely should
 - It would be ideal to create an automated test that confirms the data in the report is truly accurate
 - We either need to preserve a copy of requests made during Startup when running with "--no-reset-metrics" so we can show all data in the chart, or not show that section of time, as it's confusing/wrong to show it blank
 - I was able to run an hour long load test and got full data, the report file was only 117k

slashrsm commented on src/metrics.rs:
```diff
@@ -2951,6 +2973,25 @@ impl GooseAttack {
+            let graph_rps_template: String;
+            if !self.configuration.report_file.is_empty() {
+                let mut count = 0;
+                for (_path, path_metric) in self.metrics.requests.iter() {
```
I learned something new!

slashrsm commented on src/metrics.rs:
```diff
@@ -451,6 +454,22 @@ impl GooseRequestMetricAggregate {
+
+    /// Record request in requests per second metric.
+    pub(crate) fn record_requests_per_second(&mut self, second: usize) {
+        if self.requests_per_second.len() <= second {
```
Done.

slashrsm commented:
I played with the color scheme and aspect ratio and I think that graph looks better now. Starting phase is only displayed when `--no-reset-metrics` is used and stopping phase is only displayed if it lasts for long enough to actually be visible on the graph. 

## Screenshots

### With `--no-reset-metrics`
<img width="1041" alt="Screenshot 2021-11-18 at 20 53 10" src="https://user-images.githubusercontent.com/376889/142495119-2002196c-8cf3-44eb-ab7e-92f8f74242b9.png">


### Without `--no-reset-metrics`
<img width="1066" alt="Screenshot 2021-11-18 at 20 49 50" src="https://user-images.githubusercontent.com/376889/142495134-ed00e11c-1241-4620-854c-1308761ffc52.png">

jeremyandrews commented:
This is a very nice enhancement, and only affects the HTML report generated when the load test ends. Committing as is: the rest of your TODO will make great followups -- thanks!
